### PR TITLE
[feature/uploads] Upload files and pictures from the owncloud app

### PR DESCRIPTION
**Pull request by pablocarmu**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**CONTRIBUTING**](https://github.com/owncloud/ios-app/blob/master/CONTRIBUTING.md) document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: ownCloud/Client/ClientQueryViewController.swift [hunk 1]
```diff
@@ -849,3 +935,48 @@ extension ClientQueryViewController: UITableViewDragDelegate {
...
 
 }
+
+// MARK: - UIImagePickerControllerDelegate
+extension ClientQueryViewController: UIImagePickerControllerDelegate {
+
+	func imagePickerController(_ picker: UIImagePickerController, didFinishPickingMediaWithInfo info: [String : Any]) {
+
+		var name: String = "test.jpg"
+		var url: URL = URL(string: "https://demo.owncloud.org")!
```
Comment (felix-schwarz) [inline ownCloud/Client/ClientQueryViewController.swift]: Both`name` and `url` should probably be optionals…

File: ownCloud/Client/ClientQueryViewController.swift [hunk 2]
```diff
@@ -849,3 +935,48 @@ extension ClientQueryViewController: UITableViewDragDelegate {
...
+			name = movieURL.lastPathComponent
+			url = movieURL
+		}
+
+		if let imageAsset = info[UIImagePickerControllerPHAsset] as? PHAsset {
+			let resources = PHAssetResource.assetResources(for: imageAsset)
+			name = resources[0].originalFilename
+		}
+
+		upload(itemURL: url, name: name)
```
Comment (felix-schwarz) [inline ownCloud/Client/ClientQueryViewController.swift]: … and `upload(itemURL:name:)` only be called if `url` is non-nil.

File: ownCloud/Resources/Info.plist [hunk 3]
```diff
@@ -29,8 +33,14 @@
 		<key>NSAllowsArbitraryLoads</key>
 		<true/>
 	</dict>
+	<key>NSAppleMusicUsageDescription</key>
+	<string>We need this so that you can upload photos and videos from your photo library</string>
```
Comment (felix-schwarz) [inline ownCloud/Resources/Info.plist]: Alternative suggestion:
This permission is needed by the ownCloud app to upload media files to your server.

File: ownCloud/Resources/Info.plist [hunk 4]
```diff
@@ -29,8 +33,14 @@
...
 	<key>NSFaceIDUsageDescription</key>
 	<string>Unlock the app without entering the passcode.</string>
+	<key>NSPhotoLibraryAddUsageDescription</key>
+	<string>We need this so that you can upload photos and videos from your photo library</string>
```
Comment (felix-schwarz) [inline ownCloud/Resources/Info.plist]: Alternative suggestion:
This permission is needed by the ownCloud app to upload photos and videos from your photo library.

File: ownCloud/Resources/Info.plist [hunk 5]
```diff
@@ -29,8 +33,14 @@
...
+	<key>NSPhotoLibraryUsageDescription</key>
+	<string>We need this so that you can upload photos and videos from your photo library</string>
```
Comment (felix-schwarz) [inline ownCloud/Resources/Info.plist]: Alternative suggestion:
This permission is needed by the ownCloud app to upload photos and videos from your photo library.

File: ownCloud/Resources/Info.plist [hunk 6]
```diff
@@ -60,7 +70,5 @@
 		<string>UIInterfaceOrientationLandscapeLeft</string>
 		<string>UIInterfaceOrientationLandscapeRight</string>
 	</array>
-	<key>ITSAppUsesNonExemptEncryption</key>
-	<false/>
```
Comment (felix-schwarz) [inline ownCloud/Resources/Info.plist]: Why did you remove this?

**Comment by felix-schwarz** [comment]
For reasons unbeknownst to me, the dots grow large while being tapped:
screen shot 2018-11-15 at 17 09 10
screen shot 2018-11-15 at 17 08 41

File: ownCloud/Client/ClientQueryViewController.swift (hunk 1 above)
Comment (pablocarmu) [reply ownCloud/Client/ClientQueryViewController.swift]: Changed.

File: ownCloud/Client/ClientQueryViewController.swift (hunk 2 above)
Comment (pablocarmu) [reply ownCloud/Client/ClientQueryViewController.swift]: Done

File: ownCloud/Resources/Info.plist (hunk 3 above)
Comment (pablocarmu) [reply ownCloud/Resources/Info.plist]: Done.

File: ownCloud/Resources/Info.plist (hunk 4 above)
Comment (pablocarmu) [reply ownCloud/Resources/Info.plist]: Done.

File: ownCloud/Resources/Info.plist (hunk 5 above)
Comment (pablocarmu) [reply ownCloud/Resources/Info.plist]: Done.

File: ownCloud/Resources/Info.plist (hunk 6 above)
Comment (pablocarmu) [reply ownCloud/Resources/Info.plist]: I think this was removed without notice in a merge/rebase, now it's in place again.

File: ownCloud/Client/ClientQueryViewController.swift [hunk 7]
```diff
@@ -849,3 +936,52 @@ extension ClientQueryViewController: UITableViewDragDelegate {
...
+			url = movieURL
+		}
+
+		if let imageAsset = info[UIImagePickerControllerPHAsset] as? PHAsset {
+			let resources = PHAssetResource.assetResources(for: imageAsset)
+			name = resources[0].originalFilename
+		}
+
+		guard name != nil, url != nil else {
+			return
```
Comment (felix-schwarz) [inline ownCloud/Client/ClientQueryViewController.swift]: This way, `picker.dismiss()` will not be called later on. Might therefore be better to just check for nil with an `if` ahead of the `upload()` call instead, but let the `picker.dismiss()` be called regardless.

File: ownCloud/Resources/Info.plist [hunk 8]
```diff
@@ -29,8 +33,14 @@
 		<key>NSAllowsArbitraryLoads</key>
 		<true/>
 	</dict>
+	<key>NSAppleMusicUsageDescription</key>
+	<string>This permission is needed by the ownCloud app to upload media files to your server.</string>
```
Comment (felix-schwarz) [inline ownCloud/Resources/Info.plist]: Sorry for not thinking of this again, but mentioning "ownCloud app" might be a problem for branding.

"This permission is needed for uploading media files to your server." ?

File: ownCloud/Resources/Info.plist [hunk 9]
```diff
@@ -29,8 +33,14 @@
...
 	<key>NSFaceIDUsageDescription</key>
 	<string>Unlock the app without entering the passcode.</string>
+	<key>NSPhotoLibraryAddUsageDescription</key>
+	<string>This permission is needed by the ownCloud app to upload photos and videos from your photo library.</string>
```
Comment (felix-schwarz) [inline ownCloud/Resources/Info.plist]: Same here, sorry.

"This permission is needed to upload photos and videos from your photo library."?

File: ownCloud/Resources/Info.plist [hunk 10]
```diff
@@ -29,8 +33,14 @@
...
+	<key>NSPhotoLibraryUsageDescription</key>
+	<string>This permission is needed by the ownCloud app to upload photos and videos from your photo library.</string>
```
Comment (felix-schwarz) [inline ownCloud/Resources/Info.plist]: And once more… sorry, again.

"This permission is needed to upload photos and videos from your photo library."?

File: ownCloud/Resources/en.lproj/Localizable.strings [hunk 11]
```diff
@@ -80,6 +80,11 @@
 "Are you sure you want to delete this item from the server?" = "Are you sure you want to delete this item from the server?";
 "Deleting '%@'" = "Deleting '%@'";
 "Renaming to %@" = "Renaming to %@";
+"Upload from your photo library" = "Upload from your photo library";
+"ownCloud does not have permissons" = "ownCloud does not have permissons";
```
Comment (felix-schwarz) [inline ownCloud/Resources/en.lproj/Localizable.strings]: Little typo here. "permissons" vs. "permissions".

With branding in mind… maybe turn this into "Missing permissions"?

(also at line 700 of ClientQueryViewController.swift)

File: ownCloud/Resources/en.lproj/Localizable.strings [hunk 12]
```diff
@@ -80,6 +80,11 @@
...
+"This permission is needed by the ownCloud app to upload photos and videos from your photo library." = "This permission is needed by the ownCloud app to upload photos and videos from your photo library.";
```
Comment (felix-schwarz) [inline ownCloud/Resources/en.lproj/Localizable.strings]: With branding in mind… maybe turn this into "This permission is needed to upload photos and videos from your photo library."?

(also at line 700 of ClientQueryViewController.swift)

File: ownCloud/Client/ClientQueryViewController.swift [hunk 13]
```diff
@@ -645,6 +651,87 @@ class ClientQueryViewController: UITableViewController, Themeable {
...
+						OnMainThread {
+							self.present(alert, animated: true)
+						}
+					}
+				})
+			}
+		})
+
+		let locationsDirectory = UIAlertAction(title: "Upload file".localized, style: .default) { _ in
+			let documentDirectory = UIDocumentPickerViewController(documentTypes: [kUTTypeData as String], in: .import)
```
Comment (mneuwert) [inline ownCloud/Client/ClientQueryViewController.swift]: I think `documentDirectory` is confusing name here, since it is actually referring to a view controller instance

File: ownCloud/Client/ClientQueryViewController.swift [hunk 14]
```diff
@@ -645,6 +651,87 @@ class ClientQueryViewController: UITableViewController, Themeable {
...
+
+						OnMainThread {
+							self.present(alert, animated: true)
+						}
+					}
+				})
+			}
+		})
+
+		let locationsDirectory = UIAlertAction(title: "Upload file".localized, style: .default) { _ in
```
Comment (mneuwert) [inline ownCloud/Client/ClientQueryViewController.swift]: `locationsDirectory` is confusing. I would rather call it `uploadFileAction`

File: ownCloud/Client/ClientQueryViewController.swift [hunk 15]
```diff
@@ -145,6 +146,11 @@ class ClientQueryViewController: UITableViewController, Themeable {
 		self.tableView.dragDelegate = self
 		self.tableView.dropDelegate = self
 		self.tableView.dragInteractionEnabled = true
+
+		let actionsBarButton: UIBarButtonItem = UIBarButtonItem(title: "● ● ●", style: .done, target: self, action: #selector(uploadsBarButtonPressed))
+		actionsBarButton.setTitleTextAttributes([.font :UIFont.systemFont(ofSize: 10)], for: .normal)
```
Comment (mneuwert) [inline ownCloud/Client/ClientQueryViewController.swift]: I think we should start thinking about moving fonts into Theming support.

File: ownCloud/Client/ClientQueryViewController.swift [hunk 16]
```diff
@@ -645,6 +651,87 @@ class ClientQueryViewController: UITableViewController, Themeable {
 		}
 
 	}
+
+	func upload(itemURL: URL, name: String, viewDidAppearHandler: ClientActionVieDidAppearHandler? = nil, completionHandler: ClientActionCompletionHandler? = nil) {
```
[the remark this answers is not in the document]
Comment (pablocarmu) [reply ownCloud/Client/ClientQueryViewController.swift]: Good catch, just copy-paste from another action, I think even that we don't need this for the uploads because it's not presenting any screen in this step.

I'll remove it from the method definition.

File: ownCloud/Client/ClientQueryViewController.swift (hunk 13 above)
Comment (pablocarmu) [reply ownCloud/Client/ClientQueryViewController.swift]: I'll change it to `documentPickerVC`

File: ownCloud/Client/ClientQueryViewController.swift (hunk 14 above)
Comment (pablocarmu) [reply ownCloud/Client/ClientQueryViewController.swift]: I'll change it to `uploadFileAction` then.

File: ownCloud/Client/ClientQueryViewController.swift (hunk 15 above)
Comment (pablocarmu) [reply ownCloud/Client/ClientQueryViewController.swift]: I think there is no current mechanism implemented to allow fonts with Themes but I'm with you, we should code it and move all the fonts there.

File: ownCloud/Client/ClientQueryViewController.swift (hunk 13 above)
Comment (mneuwert) [reply ownCloud/Client/ClientQueryViewController.swift]: @pablocarmu I think we have put a rule into our coding guide, that abbreviations like `VC` shall not be used. Probably better to just name it `documentPickerViewController`. cc: @felix-schwarz

File: ownCloud/Client/ClientQueryViewController.swift (hunk 15 above)
Comment (mneuwert) [reply ownCloud/Client/ClientQueryViewController.swift]: @pablocarmu yeah, what do you think of creating a separate issue for this topic?

File: ownCloud/Client/ClientQueryViewController.swift (hunk 13 above)
Comment (pablocarmu) [reply ownCloud/Client/ClientQueryViewController.swift]: changed

File: ownCloud/Client/ClientQueryViewController.swift (hunk 15 above)
Comment (pablocarmu) [reply ownCloud/Client/ClientQueryViewController.swift]: Sure, I'll open it so we can discuss there.

**Comment by jesmrec** [comment]
Starting QA here...

**Comment by jesmrec** [comment]
### (1) 

I noticed that .HEIC files uploaded from camera roll, does not show the thumbnail. Probably this is not something to fix here.

**Comment by jesmrec** [comment]
### (2)

internal issue? or iOS one?

1. Select upload form Photo library
2. Select a video
3. `Choose`, in bottom bar

Current: Video starts to compress. By tapping in `Cancel` (bottom left) nothing happens
Expected: Cancel should stop the compression and return to picker. Or, `Cancel` disabled.

iOS12.0
iOS12

**Comment by jesmrec** [comment]
### (3)

1. Switch device connection off
2. Upload a pic from Photo library

Current: Bottom message status does not appear
Expected: Issue shown

nov-20-2018 13-29-30

**Comment by jesmrec** [comment]
### (4)

Also regarding offline:

1. Select a file to upload
2. Before the upload finishes, remove the device connection (upload stops)
3. Restablish the connection -> upload starts again

Current: file finishes to uploads but at the end an error happens:

nov-22-2018 13-15-11

Expected: Retry finishes correctly

**Comment by felix-schwarz** [comment]
@jesmrec 

(2) Lacking a modern device that does re-encoding in this place, I unfortunately can't test this. But from the description, it sounds like an OS bug.

(3) + (4) need to be dealt with in the scope of offline support in the SDK. Please reference these issues in the respective issue, so I can work through them.

**Comment by michaelstingl** [comment]
> (2) Lacking a modern device that does re-encoding in this place, I unfortunately can't test this. But from the description, it sounds like an OS bug.

@felix-schwarz I think re-encoding at upload is happening also with slow-mo videos (introduced with iPhone 5S). This might be related: https://github.com/owncloud/ios/issues/1040

**Comment by jesmrec** [comment]
### (5) [ERROR HANDLING]

Related with the error handling improvements, probably to be added here: https://github.com/owncloud/ios-app/issues/152

upload twice the same pic of the Photo album to the same target folder

Current: Non friendly error
Expected: friendly error

**Comment by jesmrec** [comment]
### (6)

also, SDK related

1. Select "Upload a file". 
2. Select a huge amount if files (>500)

Current: files not uploaded and finally, a crash. Any progress indicator
Expected: files are uploaded

**Comment by jesmrec** [comment]
I noticed this behaviour:

1. Upload a big file (you will have time to do other actions meanwhile)
2. While it is being uploaded, tap on the three-dot-button on the file cell and delete it

Current: file is deleted from the list but keep being uploaded
Expected: By deleting, we can guess the user does not want the file there, so the upload could be cancelled

For a safe behaviour, and avoid weird user actions, i would disable the three-dot-button till the upload is finished. What do you think about? @pablocarmu @michaelstingl @felix-schwarz

**Comment by michaelstingl** [comment]
> i would disable the three-dot-button

replace it with a spinner/progress indicator?

**Comment by jesmrec** [comment]
the progress is shown in the bottom bar. I would set a cancel option in case the user selects wrongl a huge wrong file. But this would go for a different issue as improvement.

Fix of @pablocarmu works for me. I will the issue

**Comment by jesmrec** [comment]
For me it is ready to go! great job!
